src/main/java/org/jitsi/utils/dsi/DominantSpeakerIdentification.java

```diff
@@ -431,6 +467,124 @@ private synchronized Speaker getOrCreateSpeaker(T id)
+        {
+            Speaker cur = loudest.get(i);
+            if (cur.getLastLevelChangedTime() < oldestValid)
+            {
+                logger.debug("Removing " + cur.id.toString() + ". old.");
+                loudest.remove(i);
+                continue;
+            }
+            if (cur == speaker)
+            {
+                logger.debug("Removing " + cur.id.toString() + ". same.");
+                loudest.remove(i);
+                continue;
+            }
+            ++i;
+        }
+
+        i = 0;
+        while (i < loudest.size())
+        {
+            Speaker cur = loudest.get(i);
+            if (cur.energyScore < speaker.energyScore)
+                break;
+            ++i;
+        }
+
+        if (i < numLoudestToTrack)
+        {
+            logger.debug("Adding " + speaker.id.toString() + " at position " + i + ".");
+            loudest.add(i, speaker);
+        }
+
+        while (loudest.size() > numLoudestToTrack)
+            loudest.remove(numLoudestToTrack);
+
+        if (logger.isDebugEnabled())
+        {
+            i = 0;
+            while (i < loudest.size())
+            {
+                Speaker cur = loudest.get(i);
+                logger.debug("New list: " + i + ": " + cur.id.toString() + ": " + cur.energyScore + ".");
+                ++i;
+            }
+        }
+    }
+
+    /**
+     * Returns information about an endpoint's energy levels relative
+     * to other endpoints.
+     */
+    public class SpeakerRanking {
+
+        /**
+         * Whether the endpoint is currently the dominant speaker.
+         */
+        public final boolean isDominant;
+
+        /**
+         * The endpoint's current rank by energy level.
+         */
+        public final int energyRanking;
+
+        /**
+         * The endpoint's energy score, a smoothed average of processed
+         * energy levels.
+         */
+        public final int energyScore;
+
+        /**
+         * Initializes a new <tt>SpeakerRanking</tt> instance.
+         */
+        public SpeakerRanking(boolean isDominant_, int energyRanking_, int energyScore_) {
+            isDominant = isDominant_;
+            energyRanking = energyRanking_;
+            energyScore = energyScore_;
+        }
+    }
+
+    /**
+     * Get current energy rank and related data for an endpoint.
+     */
+    public synchronized SpeakerRanking getRanking(T id)
+    {
+        boolean isDominant = dominantId != null && dominantId.equals(id);
+        int rank = 0;
+        while (rank < loudest.size())
+        {
+            Speaker speaker = loudest.get(rank);
+            if (speaker.id.equals(id))
+            {
+                return new SpeakerRanking(isDominant, rank, speaker.energyScore);
+            }
+            ++rank;
+        }
+
+        Speaker speaker = getOrCreateSpeaker(id);
+        int energyScore = speaker != null ? speaker.energyScore : 0;
+        return new SpeakerRanking(isDominant, rank, energyScore);
```

Comment:
This case returns `rank = loudest.size()`, even though the exact rank is unknown. Should it return `rank = -1` to make it more obvious?

Reply:
could do that. caller would need to check for the special value, though. i was also thinking this would return a valid rank in the case where the list was not at max size but the endpoint wasn't found. not sure if that could happen - maybe if all the packets from an endpoint were filtered.

Reply:
After further thought, I think this is okay. I interpret it to mean all untracked endpoints are tied for the next rank after the tracked ones. It avoids the caller needing to check for the special value and the danger of -1 being read as a higher rank than 0 if they don't check.
Added commentary to the docs explaining the return value in this case.